src/ContainerTrait.php

```diff
@@ -14,22 +39,152 @@ public function _unique_element($desired = null)
+
+            // ask element on his preferred name, then make it unique.
+            $cn = $element->getDesiredName();
+            $args[0] = $this->_unique_element($cn);
+        } else {
+
+            // generate name based on the class
+            $cn = str_replace('\\', '_', strtolower(get_class($element)));
+            $args[0] = $this->_unique_element($cn);
+        }
+
+        // Maybe element already exists
+        if (isset($this->elements[$args[0]])) {
+            throw new Exception([
+                'Element with requested name already exists',
+                'element'=>$element,
+                'name'=>$args[0],
+                'this'=>$this,
+                'arg2'=>$args
+            ]);
+        }
+
+        if(isset($this->_appScope) && isset($element->_appScope)) {
+            $element->app = $this->app;
+        }
+
+        $element->owner = $this;
+        $element->short_name = $args[0];
+
+
+        if(isset($element->_trackableTrait)) {
+
+            $element->name = $this->_shorten($this->name.'_'.$element->short_name);
+
+            $this->elements[$element->short_name] = $element;
+        } else {
+            // dont store extra reference to models and controlers
+            // for purposes of better garbage collection.
+            $this->elements[$element->short_name] = true;
+        }
+        
+        return $element;
     }
 
     /**
-     * Extension to add() method which will perform container tracking
+     * Remove child element if it exists.
+     *
+     * @param string $short_name short name of the element
+     *
+     * @return $this
      */
-    protected function _add_Container() {
-    }
+    public function removeElement($short_name)
+    {
 
-    protected function _shorten () {
+        if (is_object($short_name)) {
+            $short_name = $short_name->short_name;
+        }
+        unset($this->elements[$short_name]);
+        return $this;
     }
 
-    function getElement() {
+    /**
+     * Method used internally for shortening object names.
+     *
+     * @param string $desired Desired name of new object.
+     *
+     * @return string Shortened name of new object.
+     */
+    protected function _shorten ($desired) {
+        if (
+            isset($this->_appScope) &&
+            $this->app->max_name_length &&
+            strlen($desired) > $this->app->max_name_length
+        ) {
+            // $len is the amount to chomp. It must divide
+            // by max_name_length, so that we keep chomping
+            // by consistent amounts
+            $len = strlen($desired);
+            $len -= ($len % $this->app->max_name_length);
+
+            $key = substr($desired, 0, $len);
+            $rest = substr($desired, $len);
+
+            if (!isset($this->app->unique_hashes[$key])) {
+                $this->app->unique_hashes[$key] = dechex(crc32($key));
+            }
+            $desired = $this->app->unique_hashes[$key].'__'.$rest;
```

Comment:
This way we generate names longer than max_name_length I guess.
If max_name_length is 5 and strlen($desired) = 9, then resulting string will be 14 characters long!

See full test case here: http://sandbox.onlinephpfunctions.com/code/8864da26614afdfbf0425f16039e07d67743c629